File: src/main/java/com/handsome/summary/reading/extension/ArticleReading.java
package com.handsome.summary.reading.extension;

import static com.handsome.summary.reading.extension.ArticleReading.KIND;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.swagger.v3.oas.annotations.media.Schema;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import run.halo.app.core.extension.content.Constant;
import run.halo.app.extension.AbstractExtension;
import run.halo.app.extension.GVK;

@Data
@ToString(callSuper = true)
@GVK(kind = KIND, group = "summaraidgpt.lik.cc",
    version = Constant.VERSION, singular = "articlereading", plural = "articlereadings")
@EqualsAndHashCode(callSuper = true)
public class ArticleReading extends AbstractExtension {

    public static final String KIND = "ArticleReading";

    private Spec spec;

    @Data
    public static class Spec {
        @Schema(description = "Post metadata.name")
        private String postMetadataName;

        @Schema(description = "Post display title")
        private String postTitle;

        @Schema(description = "Post public URL")
        private String postUrl;

        @Schema(description = "Normalized article content SHA-256 hash")
        private String contentHash;

        @Schema(description = "Insight graph schema version")
        private Integer schemaVersion = 4;

        @Schema(description = "AI model metadata.name")
        private String modelName;

        @Schema(description = "Generated time")
        private Instant generatedAt;

        private InsightNode root = new InsightNode();

        private List<InsightNode> nodes = new ArrayList<>();

        private List<InsightEdge> edges = new ArrayList<>();
    }

    @Data
    public static class InsightNode {
        private String id;

        private String title;

        private String kind;

        private String summary;

        private SourceRange sourceRange;

        private InsightPayload payload = new InsightPayload();
    }

    @Data
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    public static class InsightPayload {
        private List<String> items = new ArrayList<>();
    }

    @Data
    public static class SourceRange {
        private Integer startParagraph;

        private Integer endParagraph;

        private String anchor;
    }

    @Data
    public static class InsightEdge {
        private String from;

        private String to;

        private String type;
    }
}

File: ui/src/api/article-reading.ts
import { axiosInstance, consoleApiClient, type ListedPost } from '@halo-dev/api-client'

const API_PREFIX = '/apis/api.summary.summaraidgpt.lik.cc/v1alpha1'

export interface ArticleReadingMetadata {
  name?: string
  creationTimestamp?: string
}

export type InsightNodeKind = 'root' | 'tl' | 'dl'
export type InsightEdgeType = 'contains' | 'expands' | 'supports' | 'explains'

export interface InsightPayload {
  items?: string[]
}

export interface SourceRange {
  startParagraph?: number
  endParagraph?: number
  anchor?: string
}

export interface InsightNode {
  id: string
  title: string
  kind: InsightNodeKind
  summary?: string
  sourceRange?: SourceRange
  payload?: InsightPayload
}

export interface InsightEdge {
  from: string
  to: string
  type: InsightEdgeType
}

export interface ArticleReadingSpec {
  postMetadataName?: string
  postTitle?: string
  postUrl?: string
  contentHash?: string
  schemaVersion?: number
  modelName?: string
  generatedAt?: string
  root?: InsightNode
  nodes?: InsightNode[]
  edges?: InsightEdge[]
}

export interface ArticleReading {
  metadata?: ArticleReadingMetadata
  spec?: ArticleReadingSpec
}

export interface ErrorResponse {
  success?: boolean
  message?: string
}

export interface SelectablePost {
  name: string
  title: string
  slug?: string
  permalink?: string
  phase?: string
  published?: boolean
  publishTime?: string
  updatedAt?: string
}

export interface SelectablePostPage {
  items: SelectablePost[]
  page: number
  size: number
  total: number
}

export interface ListSelectablePostsParams {
  page?: number
  size?: number
  keyword?: string
}

const toSelectablePost = (item: ListedPost): SelectablePost | undefined => {
  const post = item.post
  const name = post.metadata?.name
  if (!name || post.spec?.deleted) {
    return undefined
  }
  return {
    name,
    title: post.spec?.title || name,
    slug: post.spec?.slug,
    permalink: post.status?.permalink,
    phase: post.status?.phase,
    published: post.spec?.publish,
    publishTime: post.spec?.publishTime,
    updatedAt: post.status?.lastModifyTime || post.metadata?.creationTimestamp || undefined,
  }
}

const isSelectablePost = (post?: SelectablePost): post is SelectablePost => Boolean(post)

export const articleReadingApi = {
  async listSelectablePosts(params: ListSelectablePostsParams = {}) {
    const keyword = params.keyword?.trim()
    const { data } = await consoleApiClient.content.post.listPosts({
      page: params.page || 1,
      size: params.size || 10,
      keyword: keyword || undefined,
      sort: ['metadata.creationTimestamp,desc'],
    })
    return {
      items: (data.items || []).map(toSelectablePost).filter(isSelectablePost),
      page: data.page || params.page || 1,
      size: data.size || params.size || 10,
      total: data.total || 0,
    } satisfies SelectablePostPage
  },

  async get(postName: string) {
    const { data } = await axiosInstance.get<ArticleReading | ErrorResponse>(
      `${API_PREFIX}/articleReadings/${encodeURIComponent(postName)}`,
    )
    if (!('spec' in data) || !data.spec) {
      throw new Error((data as ErrorResponse).message || '洞察图谱尚未生成')
    }
    return data as ArticleReading
  },

  async generate(postName: string, refresh = true) {
    const { data } = await axiosInstance.post<ArticleReading | ErrorResponse>(
      `${API_PREFIX}/articleReadings`,
      {
        postName,
        refresh,
      },
    )
    if (!('spec' in data) || !data.spec) {
      throw new Error((data as ErrorResponse).message || '洞察图谱生成失败')
    }
    return data as ArticleReading
  },
}
